src/lib/core/media-marshaller/media-marshaller.ts

```diff
@@ -0,0 +1,218 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {Injectable} from '@angular/core';
+import {merge, Observable, Subject, Subscription} from 'rxjs';
+import {BreakPoint} from '../breakpoints/break-point';
+import {prioritySort} from '../breakpoints/breakpoint-tools';
+import {BreakPointRegistry} from '../breakpoints/break-point-registry';
+import {MatchMedia} from '../match-media/match-media';
+import {MediaChange} from '../media-change';
+
+type ValueMap = Map<string, string>;
+type BreakpointMap = Map<string, ValueMap>;
+type ElementMap = Map<HTMLElement, BreakpointMap>;
+type SubscriptionMap = Map<string, Subscription>;
+type WatcherMap = WeakMap<HTMLElement, SubscriptionMap>;
+type BuilderMap = WeakMap<HTMLElement, Map<string, Function>>;
+type SubjectMap = WeakMap<HTMLElement, Map<string, Subject<string>>>;
+
+/**
+ * MediaMarshaller - marshal MatchMedia events to their respective directives
+ */
+@Injectable({providedIn: 'root'})
+export class MediaMarshaller {
+  private activatedBreakpoints: BreakPoint[] = [];
+  private elementMap: ElementMap = new Map();
+  private watcherMap: WatcherMap = new WeakMap();
+  private builderMap: BuilderMap = new WeakMap();
+  private subjectMap: SubjectMap = new WeakMap();
+
+  get activatedBreakpoint(): string {
+    return this.activatedBreakpoints[0] ? this.activatedBreakpoints[0].alias : '';
+  }
+
+  constructor(protected matchMedia: MatchMedia,
+              protected breakpoints: BreakPointRegistry) {
+    this.matchMedia.observe().subscribe(this.activate.bind(this));
+  }
+
+  /**
+   * activate or deactivate a given breakpoint
+   * @param mc
+   */
+  activate(mc: MediaChange) {
+    const bp: BreakPoint | null = this.findByQuery(mc.mediaQuery);
+    if (mc.matches && bp) {
+      this.activatedBreakpoints.push(bp);
+      this.activatedBreakpoints.sort(prioritySort);
+    } else if (!mc.matches && bp) {
+      this.activatedBreakpoints.splice(this.activatedBreakpoints.indexOf(bp), 1);
+    }
+    this.updateStyles();
+  }
+
+  /**
+   * initialize the marshaller with necessary elements for delegation on an element
+   * @param element
+   * @param key
+   * @param builder optional so that custom bp directives don't have to re-provide this
+   * @param observables
+   */
+  init(element: HTMLElement,
+       key: string,
+       builder?: Function,
```

Comment:
Should we type this as a StyleBuilderFactory ?

Reply:
`Factory` implies that it returns a `StyleBuilder`, which is not the case here. I've aliased `Builder` to `Function`